### PR TITLE
Classmap tweak 3

### PR DESCRIPTION
**Pull request by lachrist**
Implemented the changes discussed in audio meeting with @brikelly, @kgilpin, and @dustinbyrne. Could not prevent myself from refactoring the entire classmap processing. The heavy changes are contained in `components/trace/appmap/classmap` in commit a72a4f210adcf915bbefb27f938e45beb49969ef. This part of the code was outdated and hard to modify. I think it is in a much better state now. In particular, the lookup of source code location has been made more robust by keeping the ast around instead of a rigid hashing function locations.

### CODE REVIEW

**Comment by lachrist** [comment]
Appmapping the juice-shop before and after this PR:

[comparison.zip](https://github.com/getappmap/appmap-agent-js/files/10041017/comparison.zip)

**Review by kgilpin — CHANGES_REQUESTED** [review]
When a value (like `class` or `value`) is "null" or "undefined", it can simply be omitted. 

```
      "receiver": {
        "name": "this",
        "class": "undefined",
        "object_id": null,
        "value": "undefined"
      },
```

Is there no `object_id` in JS? We need a way to identify when function calls are being performed on the same object.

**Comment by kgilpin** [comment]
`null` property values should be omitted. Example:

```
        "normalized_path_info": null,
```

**Comment by kgilpin** [comment]
Parameter name should be the name only; do not include type information in the name:

```
        {
          "name": "criteria: () => any",
          "class": "Function",
          "object_id": 1652,
          "value": "(...) => { ... }"
        },
```

**Comment by kgilpin** [comment]
Sometimes the agent is populating `receiver.properties` and sometimes it isn't. Can you describe how this logic works? An example:

```
      "receiver": {
        "name": "this",
        "class": "Object",
        "object_id": 430,
        "value": "[object Object]",
        "size": 9,
        "properties": [
          {
            "name": "__esModule",
            "class": "Boolean"
          },
          {
            "name": "solveIf",
            "class": "Function"
          },
          {
            "name": "solve",
            "class": "Function"
          },
          {
            "name": "sendNotification",
            "class": "Function"
          },
          {
            "name": "notSolved",
            "class": "Function"
          },
          {
            "name": "findChallengeByName",
            "class": "Function"
          },
          {
            "name": "findChallengeById",
            "class": "Function"
          },
          {
            "name": "solveFindIt",
            "class": "AsyncFunction"
          },
          {
            "name": "solveFixIt",
            "class": "AsyncFunction"
          }
        ]
      },
```

**Comment by lachrist** [comment]
> parameter.name should simply be the parameter name, it should not include type information. Type information is provided in parameter.class. Example:
>
> ```
>   "parameters": [
>         {
>           "name": "products: Product[]",
>           "class": "Array",
>           "object_id": 48,
>           "value": "[object Object],[object Object],[object Object],[object Object],[object Object],[object Object], ...",
>           "size": 44
>         }
>       ]
> ```

Currently, `parameter.name` is static information inferred from source code and `parameter.class` is dynamic information inferred from runtime values. They are complementary information. Coming up with a sound mixture of the two in `parameter.class` is going to take some work and should probably be done in separate PR.

**Comment by lachrist** [comment]
> null property values should be omitted. Example:
> ```
> "normalized_path_info": null,
> "exceptions": null
> "return_value": null
> ```

Done in 5537f07bd91f20e8a51a37aa460cc21ced4f3916

**Comment by lachrist** [comment]
> Parameter name should be the name only; do not include type information in the name:
> 
> ```
>         {
>           "name": "criteria: () => any",
>           "class": "Function",
>           "object_id": 1652,
>           "value": "(...) => { ... }"
>         },
> ```

cf comment above, I propose to disantangle types in another PR.

**Comment by lachrist** [comment]
> Sometimes the agent is populating receiver.properties and sometimes it isn't. Can you describe how this logic works?

The choice of populating properties has nothing to do with being in `receiver` position -- ie: `this`. Instead, the agent decide whether the runtime value looks like an hash. If this is the case, it will provide `properties`. The logic is in [getSpecific](https://github.com/getappmap/appmap-agent-js/blob/e7d107851596a1695b36e1e39844b278f8eec299/components/serialization/default/index.mjs#L222). Currently, if the prototype is `Object.prototype` or `null`, the agent will treat the object as an hash. It is probably a good idea to extend this to `Map`.

File: components/trace/appmap/classmap/index.mjs [hunk 1]
```diff
@@ -1,313 +1,105 @@
...
+export const lookupClassmapClosure = ({ sources }, location) => {
+  const base = getLocationBase(location);
+  if (sources.has(base)) {
+    return lookupSourceClosure(
+      sources.get(base),
+      getLocationPosition(location),
+    );
+  } else {
+    logWarning(
+      "Missing source file for closure at %s, threating it as excluded.",
```
Comment (dividedmind) [inline components/trace/appmap/classmap/index.mjs:58]: "treating" (or "threatening", but I doubt that's what you meant ;) )

**Review by dividedmind — APPROVED** [review]
You said it wasn't a behemoth ;)

File: components/error/default/index.mjs [hunk 2]
```diff
@@ -32,9 +32,8 @@ const external_message = `
 
 const unknown_message = `
 [appmap@${version}] Detected an unknown error.\
- Does this error disappear when not recording your application?\
- If yes, this is probably an issue within the appmap framework.\
- Please consider submitting a bug report at:
+ If this error disapear when not recording your application, it is probably an issue within the appmap framework.\
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/error/default/index.mjs:35]: Done 67368b3cd0790260f3be097ffd7693789c0c8559

File: components/source/default/index.mjs [hunk 3]
```diff
@@ -67,67 +93,43 @@ export const createSourceMap = ({ url: base, content }) => {
...
-            source_index += segment[1];
-            src_line += segment[2];
-            src_column += segment[3];
+    if (line <= mapping.lines.length) {
+      for (const fields of mapping.lines[line - 1]) {
+        if (fields[0] === column && fields.length >= 4) {
+          if (fields[1] < mapping.sources.length) {
+            return makeLocation(mapping.sources[fields[1]].url, {
+              line: fields[2] + 1,
+              column: fields[3],
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/source/default/index.mjs:114]: The column field? I left it zero-based as provided by locations objects from parsers.

File: components/trace/appmap/classmap/index.mjs (hunk 1 above)
Comment (lachrist) [reply components/trace/appmap/classmap/index.mjs:58]: This closure is completely out of line. I'm *threatening* of exclusion! Rebased in 3c65c2448320c276ef2d297b3d39461c5848040c

File: components/location/default/index.test.mjs [hunk 4]
```diff
@@ -1,19 +1,14 @@
-import { assertDeepEqual, assertEqual } from "../../__fixture__.mjs";
+import { assertEqual, assertDeepEqual } from "../../__fixture__.mjs";
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/location/default/index.test.mjs:1]: Yeah this was not intended :s

**Comment by lachrist** [comment]
@dividedmind thanks for the review!

**Pull request merged**